app/lib/history/models.dart

```diff
@@ -122,46 +128,72 @@ abstract class HistoryData {
   DateTime get timestamp;
 }
 
+// ignore: one_member_abstracts
 abstract class HistoryEvent {
-  String getType();
-
-  Map<String, dynamic> toJson();
-
   String formatMarkdown(HistoryData data);
 }
 
-typedef HistoryEvent HistoryEventFromJson(Map<String, dynamic> json);
+@JsonSerializable()
+class HistoryUnion {
+  final PackageVersionUploaded packageVersionUploaded;
+  final UploaderChanged uploaderChanged;
+  final AnalysisCompleted analysisCompleted;
+
+  HistoryUnion({
+    this.packageVersionUploaded,
+    this.uploaderChanged,
+    this.analysisCompleted,
+  }) {
+    assert(_items.where((x) => x != null).length == 1);
+  }
+
+  factory HistoryUnion.ofEvent(HistoryEvent event) {
+    if (event is PackageVersionUploaded) {
+      return new HistoryUnion(packageVersionUploaded: event);
+    } else if (event is UploaderChanged) {
+      return new HistoryUnion(uploaderChanged: event);
+    } else if (event is AnalysisCompleted) {
+      return new HistoryUnion(analysisCompleted: event);
+    } else {
+      throw new ArgumentError('Unknown type: ${event.runtimeType}');
+    }
+  }
```

Comment:
Couldn't we just do:
```dart
HistoryUnion.fromPackageVersionUploaded(this.PackageVersionUploaded);
HistoryUnion.fromUploaderChanged(this.UploaderChanged);
HistoryUnion.fromAnalysisCompleted(this.AnalysisCompleted);
```

Defining a one-line constructor for each option... Or maybe we could just use the first constructor, since it has an assert?

Reply:
Ah, I now see how you're using it,.. hmm... thinking...